### PR TITLE
feat(Core/DB/Creature): Throw error if npcflag and gossip_menu_id is not c…

### PR DESCRIPTION
**Pull request by Kitzunu**
…ombined

*cherry-pick commit (https://github.com/TrinityCore/TrinityCore/commit/a513922fc1572137c4e8b7b75d7ba931b15cf26a)

## Changes Proposed:
- Make sure that creatures with a gossip menu also has gossip flag
-  Make sure that creatures that does not have a gossip does not have a gossip flag+

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

**Comment by Kitzunu** [comment]
I just want to throw it out there that the module error
```
Creature (Entry: 190010) has npcflag UNIT_NPC_FLAG_GOSSIP (1), but gossip menu is unassigned.
```

Is caused by some unknown module and is not something from the core.
If someone knows which module it is, I can fix it there.

**Comment by TheLegendaryKingdom** [comment]
> I just want to throw it out there that the module error
> 
> ```
> Creature (Entry: 190010) has npcflag UNIT_NPC_FLAG_GOSSIP (1), but gossip menu is unassigned.
> ```
> 
> Is caused by some unknown module and is not something from the core. If someone knows which module it is, I can fix it there.

I guess 190010 is the [default Transmogrifier provided](https://github.com/azerothcore/mod-transmog/blob/master/data/sql/db-world/trasm_world_NPC.sql) with the module [mod-transmog](https://github.com/azerothcore/mod-transmog)

**Review by acidmanifesto — APPROVED** [review]

**Comment by acidmanifesto** [comment]
What is the Status of this PR? @Kitzunu

**Comment by Kitzunu** [comment]
trying to get a half good way to exclude entries

**Review by acidmanifesto — APPROVED** [review]

File: src/server/game/Globals/ObjectMgr.cpp [hunk 1]
```diff
@@ -1160,6 +1160,22 @@ void ObjectMgr::CheckCreatureTemplate(CreatureTemplate const* cInfo)
...
+        break;
+    default:
+        if (cInfo->GossipMenuId && !(cInfo->npcflag & UNIT_NPC_FLAG_GOSSIP))
+        {
+            LOG_ERROR("sql.sql", "Creature (Entry: %u) has assigned gossip menu %u, but npcflag does not include UNIT_NPC_FLAG_GOSSIP (1).", cInfo->Entry, cInfo->GossipMenuId);
+        }
+        else if (!cInfo->GossipMenuId && (cInfo->npcflag & UNIT_NPC_FLAG_GOSSIP))
+        {
+            LOG_ERROR("sql.sql", "Creature (Entry: %u) has npcflag UNIT_NPC_FLAG_GOSSIP (1), but gossip menu is unassigned.", cInfo->Entry);
+        }
```
Comment (FrancescoBorzi) [inline src/server/game/Globals/ObjectMgr.cpp:1177]:
```suggestion
        case 190010: // Transmog Module
            break;
        default:
            if (cInfo->GossipMenuId && !(cInfo->npcflag & UNIT_NPC_FLAG_GOSSIP))
            {
                LOG_ERROR("sql.sql", "Creature (Entry: %u) has assigned gossip menu %u, but npcflag does not include UNIT_NPC_FLAG_GOSSIP (1).", cInfo->Entry, cInfo->GossipMenuId);
            }
            else if (!cInfo->GossipMenuId && (cInfo->npcflag & UNIT_NPC_FLAG_GOSSIP))
            {
                LOG_ERROR("sql.sql", "Creature (Entry: %u) has npcflag UNIT_NPC_FLAG_GOSSIP (1), but gossip menu is unassigned.", cInfo->Entry);
            }
```

**Review by FrancescoBorzi — CHANGES_REQUESTED** [review]

File: src/server/game/Globals/ObjectMgr.cpp [hunk 2]
```diff
@@ -1160,6 +1160,22 @@ void ObjectMgr::CheckCreatureTemplate(CreatureTemplate const* cInfo)
...
+    }
```
Comment (FrancescoBorzi) [inline src/server/game/Globals/ObjectMgr.cpp:1178]:
```suggestion
    switch (cInfo->Entry)
    {
        case 190010: // Transmog Module
          return;
    }
    
    if (cInfo->GossipMenuId && !(cInfo->npcflag & UNIT_NPC_FLAG_GOSSIP))
    {
        LOG_ERROR("sql.sql", "Creature (Entry: %u) has assigned gossip menu %u, but npcflag does not include UNIT_NPC_FLAG_GOSSIP (1).", cInfo->Entry, cInfo->GossipMenuId);
    }
    else if (!cInfo->GossipMenuId && (cInfo->npcflag & UNIT_NPC_FLAG_GOSSIP))
    {
        LOG_ERROR("sql.sql", "Creature (Entry: %u) has npcflag UNIT_NPC_FLAG_GOSSIP (1), but gossip menu is unassigned.", cInfo->Entry);
    }
```

Comment (FrancescoBorzi) [inline src/server/game/Globals/ObjectMgr.cpp:1178]:
```suggestion
    switch (cInfo->Entry)
    {
        case 190010: // Transmog Module
          return;
    }
    
    if (cInfo->GossipMenuId && !(cInfo->npcflag & UNIT_NPC_FLAG_GOSSIP))
    {
        LOG_ERROR("sql.sql", "Creature (Entry: %u) has assigned gossip menu %u, but npcflag does not include UNIT_NPC_FLAG_GOSSIP (1).", cInfo->Entry, cInfo->GossipMenuId);
    }
    else if (!cInfo->GossipMenuId && (cInfo->npcflag & UNIT_NPC_FLAG_GOSSIP))
    {
        LOG_ERROR("sql.sql", "Creature (Entry: %u) has npcflag UNIT_NPC_FLAG_GOSSIP (1), but gossip menu is unassigned.", cInfo->Entry);
    }
```

**Review by FrancescoBorzi — CHANGES_REQUESTED** [review]

**Comment by FrancescoBorzi** [comment]
@Kitzunu refactored the code to look cleaner, also don't forget to add `Core` in the commit name when touching core stuff ;)

File: src/server/game/Globals/ObjectMgr.cpp [hunk 3]
```diff
@@ -1160,6 +1160,22 @@ void ObjectMgr::CheckCreatureTemplate(CreatureTemplate const* cInfo)
...
 
     const_cast<CreatureTemplate*>(cInfo)->DamageModifier *= Creature::_GetDamageMod(cInfo->rank);
+
+    // Hack for modules
+    switch (cInfo->Entry)
+    {
+        case 190010: // Transmog Module
+          return;
+    }
+
```
Comment (Kitzunu) [inline src/server/game/Globals/ObjectMgr.cpp:1170]:
```suggestion

```

**Review by Nyeriah — APPROVED** [review]

**Pull request merged**